### PR TITLE
feat(core): Lower rate-limit and better batching in filestream

### PR DESCRIPTION
**Pull request by timoffex**
Description
---
Send at most one filestream request per 15 seconds and accumulate data while waiting.

I broke out `TransmitLoop` and `CollectLoop` out of `filestreamimpl.go` because testing it thoroughly was becoming a pain.

Fixes WB-19169 (but see note below).

This solves two problems:

1. We're less likely to overload the backend with too many small requests
2. On a slow network, we'll send fewer, larger requests
    * Before, we would have sent the same requests but slower
    * Before, filestream could easily clog up, causing the service to not handle any new messages until requests go through

NOTE: There's no request size cap. In extreme cases, `filestream` may accumulate a request so large that it will never succeed. I plan to implement this later.

Testing
---
Manually run the following with debug logging:

```python
import time
from tqdm import trange
import wandb

with wandb.init() as run:
    for i in trange(60):
        run.log({"i": i})
        time.sleep(1)
```

Then examine the posted requests:

```
jq '
  select(.msg | contains("filestream: post request"))
    | [.time, (.request | fromjson)]
' wandb/latest-run/logs/debug-internal.log
```

The requests are spaced 15 seconds apart except for the last one.

### CODE REVIEW

**Comment by kptkin** [comment]
@timoffex re: cap size do we track it or you on top of it?

**Comment by timoffex** [comment]
> @timoffex re: cap size do we track it or you on top of it?

Created an issue for it: https://wandb.atlassian.net/browse/WB-19414

File: core/pkg/server/stream_init.go [hunk 1]
```diff
@@ -119,10 +120,11 @@ func NewFileStream(
...
 	params := filestream.FileStreamParams{
-		Settings:  settings.Proto,
-		Logger:    logger,
-		Printer:   printer,
-		ApiClient: fileStreamRetryClient,
+		Settings:          settings.Proto,
+		Logger:            logger,
+		Printer:           printer,
+		ApiClient:         fileStreamRetryClient,
+		TransmitRateLimit: rate.NewLimiter(rate.Every(15*time.Second), 1),
```
[the remark this answers is not in the document]
Comment (timoffex) [reply core/pkg/server/stream_init.go:127]: my test of whether something needs to be a const is:

* does it have a name? (if not, make it a named const)
* is it duplicated? (if yes, factor out to a reusable const)

in this case, it has a name (`TransmitRateLimit`, plus `rate.Every(15*time.Second)` reads well in English: "rate: every 15 seconds"), and it's only used once, so I wouldn't make a const

**Comment by timoffex** [comment]
> I saw that you removed all the wait groups, do you relay on the fact that we pass around the channels for sync?

Did this in an earlier refactor---it lets us avoid the interleaved close-wait-close-wait pattern we had in `filestream.Close`. Only the feedback WaitGroup matters: when the feedback goroutine completes, it means all updates were processed.

All the intermediate goroutines use range-loops and channel closing to communicate when work is done.

**Pull request merged**